File: QID-2273-SFEbsprices/SFEbsprices.R

# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

bs_call = function(S, K, r, tau, sigma) {
    # Black-Scholes formula for European call price with b = r (costs of carry =
    # risk free interest rate -> the underlying pays no continuous dividend)
    d1 = (log(S/K) + (r + (sigma^2/2)) * tau)/(sigma * sqrt(tau))
    d2 = d1 - (sigma * sqrt(tau))
    C  = S * pnorm(d1) - K * exp(-r * tau) * pnorm(d2)
    return(C)
}

# Input
S      = 100       # stock price
K      = 100       # strike price
r      = 0.1       # risk free interest rate
tau    = 0.6       # time to maturity (blue line)
tau2   = 0.003     # time to maturity (red dotted line)
sigma  = 0.15      # volatility of first plot
sigma2 = 0.3       # volatility of second plot
lb     = 0         # lower limit of x axis 
ub     = 170       # upper limit of x axis

# Calculate Black Scholes prices as a function of S with two different
# volatilities and time to maturity
call1   = bs_call(c(lb:ub), K, r, tau, sigma)
call1_2 = bs_call(c(lb:ub), K, r, tau2, sigma)
call2   = bs_call(c(lb:ub), K, r, tau, sigma2)
call2_2 = bs_call(c(lb:ub), K, r, tau2, sigma2)

par(mfrow = c(1, 2))  # two plots in one graphic
# Plot the Black Scholes call price (blue line) with volatility sigma and time
# to maturity tau
plot(x = c(lb:ub), y = call1, main = "Black-Scholes price, sigma=0.15", xlab = "S", ylab = expression(paste("C(S,", tau, 
    ")")), xlim = c(lb, ub), ylim = c(0, max(call1)), type = "l", col = "blue", 
    lwd = 2)
# Plot the Black Scholes call price (red dotted line) with volatility sigma and
# time to maturity tau2
lines(x = c(lb:ub), y = call1_2, type = "l", col = "red", lwd = 2, lty = 2)

# Plot the Black Scholes call price (blue line) with volatility sigma2 and time
# to maturity tau
plot(x = c(lb:ub), y = call2, main = "Black-Scholes price, sigma=0.3", xlab = "S", ylab = expression(paste("C(S,", tau, 
    ")")), xlim = c(lb, ub), ylim = c(0, max(call2)), type = "l", col = "blue", 
    lwd = 2)
# Plot the Black Scholes call price (red dotted line) with volatility sigma2
# and time to maturity tau2
lines(x = c(lb:ub), y = call2_2, type = "l", col = "red", lwd = 2, lty = 2)

File: QID-2273-SFEbsprices/SFEbsprices.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

def bs_call(S, K, r, tau, sigma):
    """
    Calculates the Black-Scholes formula for a European call option price.
    """
    # To prevent division by zero or log of zero, S is treated as a small positive number if it's 0.
    S = np.maximum(S, 1e-10)
    
    d1 = (np.log(S / K) + (r + (sigma ** 2 / 2)) * tau) / (sigma * np.sqrt(tau))
    d2 = d1 - sigma * np.sqrt(tau)
    C = S * norm.cdf(d1) - K * np.exp(-r * tau) * norm.cdf(d2)
    return C

# Input parameters
lb = 0         # Lower limit of x-axis
ub = 170       # Upper limit of x-axis
S = np.arange(lb, ub + 1)  # Stock price range

K = 100       # Strike price
r = 0.1       # Risk-free interest rate
tau = 0.6       # Time to maturity (blue line)
tau2 = 0.003     # Time to maturity (red dotted line)
sigma = 0.15      # Volatility for the first plot
sigma2 = 0.3       # Volatility for the second plot

# Calculate Black-Scholes prices for the different scenarios
call1 = bs_call(S, K, r, tau, sigma)
call1_2 = bs_call(S, K, r, tau2, sigma)
call2 = bs_call(S, K, r, tau, sigma2)
call2_2 = bs_call(S, K, r, tau2, sigma2)

# Set up a figure with two subplots, similar to R's par(mfrow=c(1, 2))
fig, axs = plt.subplots(1, 2, figsize=(14, 6))

# --- Plot 1: sigma = 0.15 ---
axs[0].plot(S, call1, color="blue", linewidth=2, label=f"Time to Maturity (τ) = {tau}")
axs[0].plot(S, call1_2, color="red", linewidth=2, linestyle="--", label=f"Time to Maturity (τ) = {tau2}")
axs[0].set_title("Black-Scholes Price (σ = 0.15)")
axs[0].set_xlabel("Stock Price (S)")
axs[0].set_ylabel("Call Price C(S, τ)")
axs[0].set_xlim(lb, ub)
axs[0].set_ylim(0, max(call1) * 1.05) # Add 5% padding to y-axis
axs[0].legend()

# --- Plot 2: sigma = 0.30 ---
axs[1].plot(S, call2, color="blue", linewidth=2, label=f"Time to Maturity (τ) = {tau}")
axs[1].plot(S, call2_2, color="red", linewidth=2, linestyle="--", label=f"Time to Maturity (τ) = {tau2}")
axs[1].set_title("Black-Scholes Price (σ = 0.30)")
axs[1].set_xlabel("Stock Price (S)")
axs[1].set_ylabel("Call Price C(S, τ)")
axs[1].set_xlim(lb, ub)
axs[1].set_ylim(0, max(call2) * 1.05) # Add 5% padding to y-axis
axs[1].legend()

# Display the plots
plt.tight_layout()
plt.show()
